stl/inc/ranges

```diff
@@ -6352,6 +6352,140 @@ namespace ranges {
 
         inline constexpr _Chunk_by_fn chunk_by;
     } // namespace views
+
+#ifdef __cpp_lib_ranges_to_container
+    // clang-format off
+    template <class _Range, class _Container>
+    concept _Sized_and_reservable = sized_range<_Range> && sized_range<_Container>
+        && requires(_Container& _Cont, const _Container& _CCont, const range_size_t<_Container> _Count) {
+            _Cont.reserve(_Count);
+            { _CCont.capacity() } -> same_as<range_size_t<_Container>>;
+            { _CCont.max_size() } -> same_as<range_size_t<_Container>>;
+        };
+
+    template <class _Rng, class _Container, class... _Types>
+    concept _Common_constructible = common_range<_Rng> && _Cpp17_input_iterator<iterator_t<_Rng>>
+        && constructible_from<_Container, iterator_t<_Rng>, iterator_t<_Rng>, _Types...>;
+
+    template <class _Container, class _Reference>
+    concept _Can_push_back = requires(_Container& _Cont) {
+        _Cont.push_back(_STD declval<_Reference>());
+    };
+
+    template <class _Container, class _Reference>
+    concept _Can_insert_end = requires(_Container& _Cont) {
+        _Cont.insert(_Cont.end(), _STD declval<_Reference>());
+    };
+
+    template <class _Rng, class _Container, class... _Types>
+    concept _Constructible_and_insertable = constructible_from<_Container, _Types...>
+        && (_Can_push_back<_Container, range_reference_t<_Rng>>
+            || _Can_insert_end<_Container, range_reference_t<_Rng>>);
+
+    template <class _Reference, class _Container>
+    _NODISCARD constexpr auto _Container_inserter(_Container& _Cont) {
+        if constexpr (_Can_push_back<_Container, _Reference>) {
+            return back_insert_iterator{_Cont};
+        } else {
+            return insert_iterator{_Cont, _Cont.end()};
+        }
+    }
+
+    template <class _Container, input_range _Rng, class... _Types>
+        requires (!view<_Container>)
+    _NODISCARD constexpr _Container to(_Rng&& _Range, _Types&&... _Args) {
+        // clang-format on
+        constexpr bool _Ref_converts = convertible_to<range_reference_t<_Rng>, range_value_t<_Container>>;
+        if constexpr (_Ref_converts && constructible_from<_Container, _Rng, _Types...>) {

```

Comment:
If we could write `if constexpr (_Ref_converts)  {} else {}` this would work, but we can't. When `_Ref_converts` is true but none of the sub-conditions are, we need to fall through into the following cases. If we write it without the `else`:
```c++
if constexpr (_Ref_converts) {
    if constexpr (...) {
        // ...
    } else if constexpr (...) {
        // ...
    }
}

if constexpr (woof) {
    // ...
} else {
    static_assert(_Always_false<...>, ...);
}
```
The `static_assert` will fire when `_Ref_converts` and one of the sub-conditions holds but `woof` does not. We could extract each of these `if constexpr` conditions into a distinct concept to avoid excess substitution, but it's hard to think it's worthwhile to add so many lines of code for a bit of throughput.

Reply:
> We could extract each of these `if constexpr` conditions into a distinct concept to avoid excess substitution, but it's hard to think it's worthwhile to add so many lines of code for a bit of throughput.

Looking again, most of these conditions correspond to custom concepts anyway. I can define two more, and make the conversion check a separate concept that each of them refine, at very little extra code cost.